overlord/snapshotstate/backend/helpers.go

```diff
@@ -111,9 +102,23 @@ func usersForUsernames(usernames []string) ([]*user.User, error) {
 	for _, username := range usernames {
 		usr, err := userLookup(username)
 		if err != nil {
-			if !isUnknownUser(err) {
-				return nil, err
-			}
+			// Treat all non-nil errors as user.Unknown{User,Group}Error's, as
+			// currently Go's handling of returned errno from get{pw,gr}nam_r
+			// in the cgo implementation of user.Lookup is lacking, and thus
+			// user.Unknown{User,Group}Error is returned only when errno is 0
+			// and the list of users/groups is empty, but as per the man page
+			// for get{pw,gr}nam_r, there are many other errno's that typical
+			// systems could return to indicate that the user/group wasn't
+			// found, however unfortunately the POSIX standard does not actually
+			// dictate what errno should be used to indicate "user/group not
+			// found", and so even if Go is more robust, it may not ever be
+			// fully robust. See from the man page:
+			//
+			// > It [POSIX.1-2001] does not call "not found" an error, hence
+			// > does not specify what value errno might have in this situation.
+			// > But that makes it impossible to recognize errors.
+			//
+			// See upstream Go issue: https://github.com/golang/go/issues/40334
 			u, e := userLookupId(username)
```

Comment:
What's the difference between `userLookup()` and `userLookupId()`? Is it intentional to call the one moments after the other?

Reply:
`userLookup` is Go's [`user.Lookup`](https://golang.org/pkg/os/user/#Lookup) (saved to a variable for mocking) that takes the username i.e. `setharnold` as the argument, while `userLookupId` is Go's [`user.LookupId`](https://golang.org/pkg/os/user/#LookupId), which takes the uid as the argument, i.e. 1000. 

It is intentional to call them both like this yes (but note not always, only if the first username call fails). The reason to call them both like this is that here the `username` variable came from user input a la `snap save --users=user1,user2,1000,12345` where each of the arguments could be either uid or a username, so we aren't sure which is which.